nixos/system/boot/bootloaders: init

RaitoBezarius opened the pull request:
###### Description of changes

This is a NixOS module to enable "multiple bootloaders" at the same time.

They all will be installed one by one in the reverse order as they are specified to let users append new bootloaders from multiple locations.

The motivation is to build newer usecases for boot stories in NixOS.

TODO:

- [x] Assert there must be at least one bootloader enabled **or** container boot is enabled.
- [x] Add documentation on how to add a new bootloader in NixOS.
- [x] Ensure no "false positives" deprecation notices appears due to poking the structure.
- [x] Ensure containers works:

```
$ nix-build -A nixosTests.containers
/nix/store/56p2i90p8wsnr28mspdbr4zp1jryvr65-vm-test-run-containers-bridge
/nix/store/y3yzsjrr1351hy4vasb1hza26421c8k9-vm-test-run-containers-ephemeral
/nix/store/d0kp2qdsjm9jl294qr6mh54ckzj7jndy-vm-test-run-containers-extra_veth
/nix/store/903z3nak787mlg1flrd3v27ascslplry-vm-test-run-containers-hosts
/nix/store/gw3nbz5880wxgxr3wdrrh95346xb3bpa-vm-test-run-containers-imperative
/nix/store/0vw7g6xh7a8m9ds9dngwz5izwdnwj80x-vm-test-run-containers-ipv4-ipv6
/nix/store/xzp9qf6lf4cl7wz5rjblj5w9b6r3icr1-vm-test-run-containers-macvlans
/nix/store/zphqr2ly1zb9n47d6lcccyy19f2qhn2v-vm-test-run-containers-names
/nix/store/5ybbm516hc3jf0x3z7ssamr9lypa6hvf-vm-test-run-nested
/nix/store/1z01ilf1wp6hfz0zzvk9ydmzq71l225g-vm-test-run-containers-physical_interfaces
/nix/store/x67p97z5mlq72zp4k49s8b1b3bqsk659-vm-test-run-containers-portforward
/nix/store/mjaqxsxjdq4mahnb2jra5si17dckycrm-vm-test-run-containers-reloadable
/nix/store/sbr46a7wvbklpd4s7n2q5zf60z3karfi-vm-test-run-containers-restart_networking
/nix/store/4m7n16wapc4gi8cirn5a2hskz5jnq7ql-vm-test-run-containers-tmpfs
/nix/store/7ia8nsh5cydydgd99pbmkadk241hm6k7-vm-test-run-containers-unified-hierarchy
```
- [x] Validate no regression in NixOS testing.

For a further PR:
- [ ] More asserts!
- [ ] Deprecate `boot.loader.efi` into `boot.efi`

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

RaitoBezarius committed:
nixos/system/boot/bootloaders: init

This is a NixOS module to enable "multiple bootloaders" at the same time.

They all will be installed one by one in the reverse order as they are specified to
let users append new bootloaders from multiple locations.

RaitoBezarius committed:
nixos/system/boot/bootloaders: init

This is a NixOS module to enable "multiple bootloaders" at the same time.

They all will be installed one by one in the reverse order as they are specified to
let users append new bootloaders from multiple locations.

samueldr commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:
```diff
@@ -82,6 +82,14 @@ in {
+      type = types.package;
+      description = lib.mdDoc ''
+        The final builder script to install this bootloader.
+        '';
```
```suggestion
      '';
```

RaitoBezarius replied:
thank you

samueldr commented on nixos/modules/system/boot/loader/bootloaders.nix:
```diff
@@ -0,0 +1,66 @@
+      '';
+    };
+    boot.loader.enabled = mkOption {
+      type = types.enum [
+        "systemd-boot"
+        "grub"
+        "generic-extlinux-compatible"
+      ];
```
I don't know what the evaluation performance implications are, but I have a scheme in use for Mobile NixOS that makes "subnamespaces" self-describing with enums.

 - https://github.com/NixOS/mobile-nixos/blob/ba4638836e94a8f16d1d1f9e8c0530b86078029c/modules/system-types.nix#L34
 - https://github.com/NixOS/mobile-nixos/blob/ba4638836e94a8f16d1d1f9e8c0530b86078029c/modules/system-types.nix#L26-L28

And e.g.

 - https://github.com/NixOS/mobile-nixos/blob/ba4638836e94a8f16d1d1f9e8c0530b86078029c/modules/system-types/depthcharge/default.nix#L55

Such a scheme would be *especially* useful to allow external bootloader interfaces to be added without modifying NixOS proper.

RaitoBezarius replied:
I think it would be helpful to have @roberth from the getgo (and FWIW, if it was not clear, this PR has no intention to break things and backward compatibility yet, I am not even expecting to land it in 23.11 FWIW.)

RaitoBezarius replied:
OK, I see better for the enums, interesting.

samueldr replied:
It might even be possible without an intermediary option, because of the way `boot.loader.builders` is made:

```
default = type.enum builtins.attrNames boot.loader.builders
```

RaitoBezarius replied:
This will break doc evaluation

roberth replied:
Is the point to pick a single bootloader?
In that case I'd do something like samueldr's suggestion, but using option type merging instead.

`loader.nix`
```nix
options.loader = mkOption {
  type = lib.types.enum [];
  description = ''
    Which bootloader to install.
  '';
}
```

`grub.nix`
```nix
options.loader = mkOption {
  type = lib.types.enum ["grub"];
}
```

`systemd-boot.nix`
```nix
options.loader = mkOption {
  type = lib.types.enum ["systemd-boot"];
}
```

When all three modules are loaded, the types are merged such that the enum is allows the union of all values.

> This will break doc evaluation

I wouldn't be sure of that. You can have some dependencies on `config` in your types and descriptions etc, unless they depend on things that aren't available during doc evaluation, such as `pkgs` in NixOS iirc. You do have to be careful not to include default values in a way that suggests they're unchangeable, but that's not a problem here I think.

RaitoBezarius replied:
This would definitely support multiple bootloaders, we went for plain subtrees as you can see in the latest, I'm not sure we need to use type merging at all in fact, what do you think of the latest?

> I wouldn't be sure of that. You can have some dependencies on config in your types and descriptions etc, unless they depend on things that aren't available during doc evaluation, such as pkgs in NixOS iirc. You do have to be careful not to include default values in a way that suggests they're unchangeable, but that's not a problem here I think.

Oh, interesting. In this situation, we let the "subtree" express the availabilities of "loader types".

> Is the point to pick a single bootloader?

We want to offer any combination of multiple bootloaders.

RaitoBezarius replied:
Currently, I'm not using option type merging but I will keep this option in mind and feel free to propose it on the current iteration of the code.

samueldr commented on nixos/modules/system/boot/loader/bootloaders.nix:
```diff
@@ -0,0 +1,66 @@
+  # Interface
+  options = {
+    boot.loader.builders = mkOption {
+      type = types.attrsOf (types.submodule ({ ... }: {
+        options = {
+          id = mkOption {
+            type = types.str;
+            description = lib.mdDoc "A string describing the bootloader name";
+          };
+          script = mkOption {
+            type = types.package;
+            description = lib.mdDoc "An executable script that ...";
+          };
+        };
+      }));
+      default = {};
+      example = literalExpression ''{ systemd-boot = finalSystemBootBuilder; }'';
+      internal = true;
+      description = lib.mdDoc ''
+        An attribute set of builder for any potential bootloader.
+        This is set internally by each bootloader with their
+        final system boot builder script.
+      '';
+    };
```
More of a stylistic concern, can we move the submodule declaration into the `let` attribute. to make the actual option easier to parse visually?

Right now it took me a small while to realise that `boot.loader.builders` did not have options `id` and `script`, but was composed of submodules.

RaitoBezarius replied:
For sure

RaitoBezarius replied:
Let me know if you feel like this is still a concern in the current model.

RaitoBezarius committed:
nixos/tests/boot/iso-media: init

This is a new boot test using the new bootloader API and modern NixOS test API but not the latest!

RaitoBezarius commented on nixos/modules/system/boot/loader/bootloaders.nix:
```diff
@@ -0,0 +1,69 @@
+      '';
+    };
+    boot.loader.enabled = mkOption {
+      type = types.listOf (types.enum [ "systemd-boot" "grub" ]); # TODO: this will break the doc evaluation? (builtins.attrNames cfg.builders));
```
I'm not sure yet if order is important for all the usecases… A bootloader may want to depend on another in the list in an opportunistic fashion for example?

RaitoBezarius replied:
Though, I feel like we should not necessarily try to concern ourselves with those problems.

RaitoBezarius replied:
Let me know if you feel like this is still relevant, resolving in the meantime because I think I removed the ordering issue.

RaitoBezarius committed:
nixos/lib/utils: add `mkBootLoaderOptions`

This is a helper to build bootloader options and avoid duplication.

RaitoBezarius committed:
nixos/lib/utils: add `enabledBootloader`

This is a helper to derive which bootloader is enabled, while avoiding
any deprecation notice or wrong structures, e.g. `boot.loader.timeout`

RaitoBezarius committed:
nixos/system/boot/loader: boot.loader.timeout -> boot.loader.defaults.timeout

We use here a fake stub module called `defaults` which is always disabled to store
data-only global settings.

RaitoBezarius committed:
nixos/virtualisation/qemu-vm: disable bootloader when not needed

RaitoBezarius committed:
nixos/system/boot/loader: introduce stub direct boot loader

This is a stub direct boot loader anytime someone needs to direct boot something, usecase: QEMU framework.

RaitoBezarius commented on nixos/modules/system/boot/systemd/initrd-secrets.nix:
```diff
@@ -1,9 +1,13 @@
 { config, pkgs, lib, ... }:
-
+let
+  enabledBootloaders = builtins.attrNames (lib.filterAttrs (name: value: value.enable) config.boot.loader);
```
This should use `utils.enabledBootloader`

RaitoBezarius commented on nixos/lib/eval-cacheable-options.nix:
```diff
@@ -18,6 +18,7 @@ let
   utils = import "${nixosPath}/lib/utils.nix" {
     inherit config lib;
     pkgs = null;
+    options = null;
```
@pennae Is that okay for what I'm doing?

RaitoBezarius replied:
answer: No

RaitoBezarius commented on nixos/lib/utils.nix:
```diff
@@ -230,4 +230,16 @@ rec {
+
+  /* This will detect if a bootloader is enabled
+   * without triggering deprecation notices     */
+  enabledBootloader = name: value:
+  let
+    option = options.boot.loader.${name}.enable or null;
+    acceptableBootloader = if option == null then false else option.definitions != [];
+  in
+  acceptableBootloader && value.enable;
```
Ah you just commented :D

RaitoBezarius replied:
OK, will do that!

RaitoBezarius commented on nixos/modules/system/boot/loader/bootloaders.nix:
```diff
@@ -0,0 +1,44 @@
+    boot.loaderId = mkOption {
+      type = types.str;
+      default = foldEnabledBootloaders (builder: previousId: "${previousId}.${builder.id}") "";
+      defaultText = ''foldEnabledBootloaders (builder: previousId: "''${previousId}-''${builder.id}") ""'';
```
Thank you!

RaitoBezarius commented on nixos/modules/system/boot/loader/defaults.nix:
```diff
@@ -0,0 +1,35 @@
+
+  options.boot.loader.defaults = utils.mkLoaderOption {
+    enable = mkOption {
+      # make me readonly
```
awesome

RaitoBezarius commented on nixos/modules/system/boot/loader/external/external.nix:
```diff
@@ -25,12 +30,6 @@ in
-    };
-
-    system.build.installBootLoader = cfg.installHook;
+    boot.loader.external.supportsInitrdSecrets = mkDefault false;
```
Hmm, external should be re-assignable, no?

RaitoBezarius replied:
Yeah you're right on that

RaitoBezarius replied:
Hmmm, I would like to resume this discussion @pennae ; how do you feel about this? I agree that uniq bool would be much better, maybe the solution is to always ask users to explicitly set false or true for their models? Not a big ask IMHO.

RaitoBezarius replied:
Logged in the referred issue, resolving in the meantime.

RaitoBezarius commented:
@grahamc @cole-h I added you because you are co-maintainers of the external loader implementation, in this PR, we will probably deprecate it as this PR achieves the same with "more" flexibility on the long run. This PR won't move too fast, so no hurry on reviews.

SuperSandro2000 commented on nixos/modules/system/boot/loader/bootloaders.md:
```diff
@@ -0,0 +1,29 @@
+
+## Installing multiple bootloaders at once {#sec-bootloaders-multiple}
+
+NixOS supports installing multiple bootloaders at once, most of the users are interested into installing one bootloader,
```
```suggestion
NixOS supports installing multiple bootloaders at once, but most users will only require one bootloader,
```

SuperSandro2000 commented on nixos/modules/system/boot/loader/bootloaders.md:
```diff
@@ -0,0 +1,29 @@
+## Installing multiple bootloaders at once {#sec-bootloaders-multiple}
+
+NixOS supports installing multiple bootloaders at once, most of the users are interested into installing one bootloader,
+and this is what the NixOS installers will often do.
```
```suggestion
and most NixOS installers will do that.
```

SuperSandro2000 commented on nixos/modules/system/boot/loader/bootloaders.md:
```diff
@@ -0,0 +1,29 @@
+NixOS supports installing multiple bootloaders at once, most of the users are interested into installing one bootloader,
+and this is what the NixOS installers will often do.
+
+Nevertheless, there are situations where it is desirable to install multiple bootloaders at once inside of a NixOS system,
```
```suggestion
Nevertheless, there are situations where it is desirable to install multiple bootloaders at once inside one system,
```

SuperSandro2000 commented on nixos/modules/system/boot/loader/bootloaders.md:
```diff
@@ -0,0 +1,29 @@
+for example:
+
+- an ISO image with GRUB as legacy bootloader and systemd-boot as UEFI bootloader.
+- an image with systemd-boot as UEFI bootloader and generic-extlinux-compatible configuration file for any firmware that does not work with UEFI by default, e.g. U-Boot based system.
```
```suggestion
- an ISO image with systemd-boot as the UEFI bootloader and generic-extlinux-compatible configuration file for any firmware that does not work with UEFI by default, like U-Boot based system.
```

RaitoBezarius replied:
Thank you for your suggestions, but please avoid changing `e.g.` in `like`, I don't see any good reasons to do so. If you care about that sort of things, consider doing your own follow-up PRs. This is just noise to me.

SuperSandro2000 commented on nixos/modules/system/boot/loader/bootloaders.nix:
```diff
@@ -0,0 +1,44 @@
+    doc = ./bootloaders.md;
+  };
+
+  # Interface
```
```suggestion
```

SuperSandro2000 commented on nixos/modules/system/boot/loader/bootloaders.nix:
```diff
@@ -0,0 +1,44 @@
+    };
+  };
+
+  # Implementation
```
```suggestion
```

SuperSandro2000 commented on nixos/modules/system/boot/loader/defaults.nix:
```diff
@@ -0,0 +1,35 @@
+      default = 5;
+      type = types.nullOr types.int;
+      description = lib.mdDoc ''
+        Timeout (in seconds) until loader boots the default menu item. Use null if the loader menu should be displayed indefinitely.
```
```suggestion
        Timeout (in seconds) until bootloader starts the default menu item. Use `null` to require the user to select an entry.
```

SuperSandro2000 commented on nixos/modules/system/boot/loader/defaults.nix:
```diff
@@ -0,0 +1,35 @@
+      description = lib.mdDoc ''
+        Timeout (in seconds) until loader boots the default menu item. Use null if the loader menu should be displayed indefinitely.
+
+        Not all bootloaders supports timeouts.
```
```suggestion
        :::NOTE
        Not all bootloaders supports timeouts.
        :::
```
Was that the syntax? Can't remember and searching on mobile is not fun.

RaitoBezarius replied:
Let me know when you find the syntax in the meantime, will resolve until then.

samueldr commented on nixos/lib/make-loader-options.nix:
```diff
@@ -0,0 +1,38 @@
+{ lib }:
```
rename file `make-bootloader-options`, and be mindful about loader vs. bootloader usage everywhere.

samueldr commented on nixos/lib/make-loader-options.nix:
```diff
@@ -0,0 +1,38 @@
+  commonOptions = {
+      id = mkOption {
+        type = types.str;
+        description = mdDoc "Identifier of this external bootloader for metadata";
```
This one should be internal too, as it is useless for end-users.

samueldr commented on nixos/lib/make-loader-options.nix:
```diff
@@ -0,0 +1,38 @@
+        internal = true;
+        description = mdDoc ''
+         Whether the bootloader setup runs append-initrd-secrets.
+         If not, any needed secrets must be copied into the initrd
+         and thus added to the store.
+
+         This means that the "secrets" are absolutely not secrets
+         with respect to that feature.
```
Hm? What do you mean here?

I think the alternative is that if it's not supported it's not supported. This sounds like it would implicitly add secrets in an insecure way.

```suggestion
         When unsupported, and initrdSecrets are configured to be used,
         the system build will error out accordingly.
```

Or am I misunderstanding?

RaitoBezarius commented on nixos/modules/system/boot/loader/bootloaders.nix:
```diff
@@ -0,0 +1,69 @@
+            description = lib.mdDoc "An executable script that ...";
+          };
+        };
+      }));
```
Yep, I moved to something closer to option type merging if needed at all.

RaitoBezarius committed:
nixos/installer/cd-dvd/iso-image: use `boot.loader.defaults.timeout`

Instead of `boot.loader.timeout`.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
lib/modules: colored warnings for obsolete options

When an option was used with a mkRenamedOptionModule, the warning was not red.

This uses `lib.warn` instead of a manual `builtins.trace`.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius commented:
@vcunat Could I request a Hydra jobset for this branch?

RaitoBezarius committed:
nixos/system/boot/bootloaders: assert at least one bootloader

This ensures the system is not dysfunctional, even in cases of "no bootloader", e.g. direct boot.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
nixos/tests/all-tests: group all container tests

This makes building them much easier for test.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius commented:
PTAL @samueldr @pennae @roberth

Gentle ping @cole-h @grahamc (or I will proceed without your feedback)

cc @ElvishJerricco @nikstur @JulienMalka as this might be of interest to you.

K900 reviewed (approved):
Like this a lot, don't really want to nitpick.

vcunat commented:
https://hydra.nixos.org/jobset/nixos/pr-244907-bootloaders-init

roberth commented on nixos/modules/system/boot/loader/bootloaders.md:
```diff
@@ -0,0 +1,112 @@
+and/or assert for its existence at evaluation time in your NixOS module.
+
+It is recommended to consider the consequences of using an installation script
+written in a language requiring a lot of dependencies, this implies that your
```
```suggestion
written in a language requiring a lot of dependencies, as this implies that your
```

roberth commented on nixos/modules/system/boot/loader/bootloaders.md:
```diff
@@ -0,0 +1,112 @@
+
+It is recommended to consider the consequences of using an installation script
+written in a language requiring a lot of dependencies, this implies that your
+users will have to download the final script's runtime closure, in cases of
+rebuild, they will need to also download the final script's build time closure.
```
```suggestion
users will have to download the final script's runtime closure, in every rebuild of their configuration they will need to also download the final script's build time closure.
```

roberth commented on nixos/lib/make-bootloader-options.nix:
```diff
@@ -0,0 +1,35 @@
+{ lib }:
+/* Build a bootloader set of options */
```
I'd prefer a module instead of module generation. Morally it definitely should, but not sure if the existing option structure and its use are receptive to a change to submodules. (Not in `attrsOf`; just submodules)

RaitoBezarius replied:
"it definitely should" → should what?
I don't really understand the rest of the sentence of the existing option structure and changes to submodules, I would appreciate you clarifying what you meant.

roberth replied:
> "it definitely should" →

be a module, morally

"Morally" does a lot of heavy lifting in that sentence, which I could be a bit confusing I guess.

Basically what I'm proposing is to change e.g. `boot.loader.systemd-boot` etc to be a submodule, each of which can individually import a module with just these options at the top level.

It does change the structure of the `options` module argument a bit, changing e.g. `options.boot.loader.systemd-boot` to an option instead of a tree of options. Not sure what's the trade-off here; whether that's a change worth making.

Let's consider this idea not a blocker.

RaitoBezarius replied:
Logged in the referred issue. Resolving in the meantime.

roberth commented on nixos/lib/make-bootloader-options.nix:
```diff
@@ -0,0 +1,35 @@
+  inherit (lib) mkOption types mdDoc;
+
+  commonOptions = {
+      id = mkOption {
```
Indentation is weird. Please run `nixpkgs-fmt` or the tool in the recent formatting RFC.

roberth reviewed (changes requested):
This is a quick review, not a complete one, esp wrt docs.

The `installBootloader` default value should be replaced by a general warning system for warnings that are displayed in the activation log. I believe this affects how the direct boot / no boot situation should be handled. This PR adds more tech debt on top of that, complicating the situation; see my comment on the code.

nitpick: The documentation team prefers a single sentence per line.

RaitoBezarius commented on nixos/modules/system/boot/loader/bootloaders.md:
```diff
@@ -0,0 +1,112 @@
+# Bootloaders in NixOS {#sec-bootloaders}
+
+NixOS has a rich and sophisticated boot story.
```
I can remove that line if you want and you can enjoy the technical terms just after that line

RaitoBezarius commented:
@roberth 

> The installBootloader default value should be replaced by a general warning system for warnings that are displayed in the activation log. I believe this affects how the direct boot / no boot situation should be handled.

I don't understand what that means, can you clarify or explain it in a different way?

RaitoBezarius commented on nixos/modules/system/boot/loader/direct-boot.nix:
```diff
@@ -0,0 +1,24 @@
+  };
+
+  options.boot.loader.directBoot = utils.mkBootLoaderOption {
+    enable = mkEnableOption (lib.mdDoc "a direct boot 'fake bootloader'");
```
Sure, we can do that.

roberth commented on nixos/modules/system/boot/loader/bootloaders.md:
```diff
@@ -0,0 +1,112 @@
+# Bootloaders in NixOS {#sec-bootloaders}
+
+NixOS has a rich and sophisticated boot story.
+
+There are multiple ways to boot NixOS:
```
```suggestion
NixOS can boot in multiple ways:
```

RaitoBezarius commented:
I really hate this baseline evaluation thingie... cd5c10f69676a36ae44254b8cfd58f528a062f65 selected for baseline.

RaitoBezarius commented:
@vcunat could I ask you to point the branch of the jobset to another branch (on nixpkgs repo) so I can have the PR branch and the jobset branch separate and easily manipulate them without closing this PR?

vcunat commented:
Sure, which branch?

RaitoBezarius commented:
> Sure, which branch?

you could create a pr-jobset-244907 branch I'd say?

vcunat commented:
Oh, I missed this.   I believe I am allowed to create branches; you are not?  I mean, creation isn't the issue really, it's more about updating the branch afterwards &ndash; it would be nice if you could do that yourself.  (I'm not immediately sure what commit you want there, too.)

RaitoBezarius commented:
> Oh, I missed this. I believe I am allowed to create branches; you are not? I mean, creation isn't the issue really, it's more about updating the branch afterwards – it would be nice if you could do that yourself. (I'm not immediately sure what commit you want there, too.)

I can update any branches and create them too.

vcunat commented:
I'm not sure I understand well, but I created [the branch](https://github.com/NixOS/nixpkgs/commit/pr-jobset-244907), merged `nixos-unstable-small` into it and switched [the jobset](https://hydra.nixos.org/jobset/nixos/pr-244907-bootloaders-init) to it.

RaitoBezarius commented:
If I know how to read well, https://hydra.nixos.org/eval/1804693 means that there are no tests regressions excluding prior ones or flakey tests.

RaitoBezarius commented:
Well, given that tests like the installers were still failing on unstable-small... https://hydra.nixos.org/eval/1804693#tabs-still-fail

Unsure if this is a good idea. I may need to rebase this on unstable I guess rather than unstable-small.
